test/metabase/api/notify_test.clj

```diff
@@ -1,28 +1,83 @@
-      (is (= (get middleware.u/response-forbidden :body)
-             (http/client :post 403 "notify/db/100"))))))
+  (with-api-key "testing-api-key"
+    (testing "POST /api/notify/db/:id"
+      (testing "endpoint should require authentication"
+        (is (= (get middleware.u/response-forbidden :body)
+               (http/client :post 403 "notify/db/100")))))))
 
 (deftest not-found-test
-  (testing "POST /api/notify/db/:id"
-    (testing "database must exist or we get a 404"
-      (is (= {:status 404
-              :body   "Not found."}
-             (try (client/post (http/build-url (format "notify/db/%d" Integer/MAX_VALUE) {})
-                               {:accept  :json
-                                :headers {"X-METABASE-APIKEY" "test-api-key"
-                                          "Content-Type"      "application/json"}})
-                  (catch clojure.lang.ExceptionInfo e
-                    (select-keys (:object (ex-data e)) [:status :body]))))))))
+  (with-api-key "testing-api-key"
+    (testing "POST /api/notify/db/:id"
+      (testing "database must exist or we get a 404"
+        (is (= {:status 404
+                :body   "Not found."}
+               (try (client/post (http/build-url (format "notify/db/%d" Integer/MAX_VALUE) {})
+                                 {:accept  :json
+                                  :headers {"X-METABASE-APIKEY" "testing-api-key"
+                                            "Content-Type"      "application/json"}})
+                    (catch clojure.lang.ExceptionInfo e
+                      (select-keys (:object (ex-data e)) [:status :body])))))))))
 
-;; TODO - how can we validate the normal scenario given that it just kicks off a background job?
+(deftest post-db-id-test
+  (with-api-key "testing-api-key"
+    (mt/test-drivers (mt/normal-drivers)
+      (let [table-name (->> (mt/db) database/tables first :name)
+            post       (fn post-api
+                         ([payload] (post-api payload 200))
+                         ([payload expected-code]
+                          (mt/client :post expected-code (format "notify/db/%d" (u/the-id (mt/db)))
+                                     {:request-options
+                                      {:headers {"X-METABASE-APIKEY" "testing-api-key"
+                                                 "Content-Type"      "application/json"}}}
+                                     payload)))]
+        (testing "sync just table when table is provided"
+          (let [long-sync-called? (atom false), short-sync-called? (atom false)]
+            (with-redefs [metabase.sync/sync-table!                        (fn [_table] (reset! long-sync-called? true))
+                          metabase.sync.sync-metadata/sync-table-metadata! (fn [_table] (reset! short-sync-called? true))]
+              (post {:scan :full, :table_name table-name})
+              (is @long-sync-called?)
+              (is (not @short-sync-called?)))))
```

Comment:
Isn't this going to be a race condition? If the sync happens asynchronously (because it's wrapped in `future`) the different sync functions aren't going to get called until some time after the API endpoint completes. I'm surprised you haven't ran in to it locally, but on CircleCI you can count of it randomly not running the other thread for 30 seconds or whatever.

I'm not sure using atoms is going to work here... you could use promises and then `u/deref-with-timeout` them with something generous (like 30 seconds for when Circle decides to be super slow), but you don't want to wait 30 seconds to find out something *didn't* run. Maybe since only one sync function ends up getting called you can use a single promise and have it get set to a value signifying which sync function is ran

Reply:
oh good point. you're right. I imagine there could be some complex stuff done to submit tasks and put an executor behind a protocol and test that or perhaps make some kind of execute function which takes a thunk and in tests just runs the thunk but in real calls `(thread (thunk))` but fundamentally this is a pretty simple `cond` that we're testing. I'll probably just remove the tests, leaving just the authenticated and 404 tests.